core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java

```diff
@@ -0,0 +1,194 @@
+/**
+ * Licensed to DigitalPebble Ltd under one or more contributor license agreements. See the NOTICE
+ * file distributed with this work for additional information regarding copyright ownership.
+ * DigitalPebble licenses this file to You under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * <p>http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * <p>Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.digitalpebble.stormcrawler.protocol.selenium;
+
+import com.digitalpebble.stormcrawler.Metadata;
+import com.digitalpebble.stormcrawler.protocol.AbstractHttpProtocol;
+import com.digitalpebble.stormcrawler.protocol.HttpHeaders;
+import com.digitalpebble.stormcrawler.protocol.ProtocolResponse;
+import com.digitalpebble.stormcrawler.util.ConfUtils;
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.LinkedBlockingQueue;
+import org.apache.storm.Config;
+import org.openqa.selenium.remote.DesiredCapabilities;
+import org.openqa.selenium.remote.RemoteWebDriver;
+import org.slf4j.LoggerFactory;
+
+public abstract class SeleniumGridProtocol extends AbstractHttpProtocol {
+    protected static final org.slf4j.Logger LOG =
+            LoggerFactory.getLogger(SeleniumGridProtocol.class);
+    protected static LinkedBlockingQueue<Holder> driversQueue;
+    private NavigationFilters filters;
+    private final DesiredCapabilities capabilities = new DesiredCapabilities();
+
+    protected String gridAddress;
+
+    @Override
+    public void configure(Config conf) {
+        super.configure(conf);
+        driversQueue = new LinkedBlockingQueue<>();
+        filters = NavigationFilters.fromConf(conf);
+        gridAddress =
+                ConfUtils.getString(conf, "selenium.grid.address", "http://localhost:4444/wd/hub");
+    }
+
+    protected synchronized List<Map<String, Object>> getAllNodesList() throws IOException {
+        Map<String, Object> valueMap = null;
+        boolean ready = false;
+        while (!ready) {
```

Comment:
I am wondering, if we should make this async and use a latch here to prevent active waiting? In any case, we might want to add some sort of wait here to prevent hammering the grid?